06-security.md

```diff
@@ -1 +1,451 @@
+
+The following example gives the CORS response headers for an OPTIONS request.
+
+```
+Access-Control-Allow-Credentials: true
+Access-Control-Allow-Origin: https://somedomain.org, https://otherdomain.org
+Access-Control-Allow-Methods: GET, OPTIONS, POST, PUT, PATCH, DELETE
+Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, X-XSRF-TOKEN
+Access-Control-Max-Age: 3600
+```
+
+Access-Control-Allow-Credentials allows the passing of a session cookie. 
+The Access-Control-Allow-Origin lists the domains that are allowed to access the API.
+The Access-Control-Allow-Methods list the permissible HTTP Methods.
+Access-Control-Allow-Headers informs the browser of the allowed headers to send.
+Finally Access-Control-Max-Age informs the browser on how long the information in the response may be cached.
+
+*Do not* use the wildcard on Access-Control-Allow-Origin unless the API must be accessible to the world.  
+```
+Access-Control-Allow-Origin: *
+``` 
+#### Use an API proxy if needed
+
+If the API does not support CORS headers you can implement a server side proxy for that service. The browser calls an URL on your server and the server invokes the API on the browsers behalf. The server does not have any SOP restrictions.
+
+Also some 3rd party APIs require a secret token to work. This token cannot be held in the single page web application because there it's easily exposed to prying eyes.
+By using a proxy this secret is easily added to the request to the 3rd party server while keeping it a secret for the browser.
+
+In order to prevent abuse, all calls to the API Proxy must be authenticated and have their XSRF token checked.
+
+#### Do not use JSONP
+
+JSONP is a SOP workaround that works by dynamically adding a script tag to the HTML document. The src of the script tag is the request to be made and the response is a JavaScript file that is immediately executed. JSONP is superseded by using CORS headers and is [difficult to do securely](http://stackoverflow.com/questions/613962/is-jsonp-safe-to-use).   
+
+### Use browser security response headers.
+
+Various [headers](https://www.owasp.org/index.php/List_of_useful_HTTP_headers) exist to prevent abuse of resources in and from the browser. 
+
+```
+X-Frame-Options: deny
+X-XSS-Protection: 1; mode=block
+X-Content-Type-Options: nosniff
+```
+
+The `X-Frame-Options` header disallows the contents to be displayed in an IFrame. The `X-XSS-Protection` header makes the browser apply a filter to the URL to prevent reflected Cross Site Scripting. Finally, the `X-Content-Type-Options` header disallows the browser from guessing the content type. 
+
+### Disable caching
+
```

Comment:
This section contradicts parts of what I write in the 'various scenarios' chapter. I think yours is right in 90%+ of the cases, and possibly 100% of non-public APIs.